apps/storefront/src/index.html

```diff
@@ -6,5 +6,72 @@
   <body>
     <oryx-app></oryx-app>
   </body>
+  <script>
+    const EVENTS_DATA = Symbol.for('EVENTS_DATA');
+    const replayableAttribute = 'replayable';
+
+    const treewalk = (
+      selector,
+      rootNode = document.body,
+      includeRoot = true
+    ) => {
+      const nodes = [rootNode];
+      const elements = [];
+
+      for (let i = 0; i < nodes.length; i++) {
+        const node = nodes[i];
+
+        if (node.nodeType !== Node.ELEMENT_NODE) {
+          continue;
+        }
+
+        if (node.children.length) {
+          nodes.push(...node.children);
+        }
+
+        if (node.shadowRoot?.children.length) {
+          nodes.push(...node.shadowRoot.children);
+        }
+
+        if (!includeRoot && node.matches(rootNode.tagName.toLowerCase())) {
+          continue;
+        }
+
+        if (node.matches(selector)) {
+          elements.push(node);
+        }
+      }
+
+      return elements;
+    };
+
+    const addEventsAction = () => {
+      const elements = treewalk(`[${replayableAttribute}]`);
+
+      for (const element of elements) {
+        element[EVENTS_DATA] = {
+          events: [],
+          listener: (event) => {
+            event.stopPropagation();
+            event.preventDefault();
+
+            element[EVENTS_DATA]?.events.push(event);
+          },
+        };
+
+        const events = element.getAttribute(replayableAttribute);
+
+        for (const eventType of events.split(':')) {
+          element.addEventListener(eventType, element[EVENTS_DATA].listener);
+        }
+      }
+    };
+
+    const isShadow = !!document.querySelector?.('oryx-app').shadowRoot;
```

Comment:
How do we go about custom root elements?

Reply:
When we move it to module, theoretically we should be able to drop it